test/Common/MessageSorterTests.cs

```diff
@@ -0,0 +1,316 @@
+        }
+
+        [Fact]
+        [Trait("Category", PlatformSpecificHelpers.TestCategory)]
+        public void Collection()
+        {
+            var senderId = "A";
+            var receiverId = "B";
+
+            var senderSorter = new MessageSorter();
+
+            var message1 = Send(senderId, receiverId, "1", senderSorter, DateTime.UtcNow);
+            var message2 = Send(senderId, receiverId, "2", senderSorter, DateTime.UtcNow + ReorderWindow + ReorderWindow);
+            var message3 = Send(senderId, receiverId, "3", senderSorter, DateTime.UtcNow + ReorderWindow + ReorderWindow + ReorderWindow + ReorderWindow);
+
+            List<RequestMessage> batch;
+            MessageSorter receiverSorter = new MessageSorter();
+
+            // delivering the sequence in order produces 1 message each time
+            receiverSorter = new MessageSorter();
+            batch = receiverSorter.ReceiveInOrder(message1, ReorderWindow).ToList();
+            Assert.Single(batch).Input.Equals("1");
+            batch = receiverSorter.ReceiveInOrder(message2, ReorderWindow).ToList();
+            Assert.Single(batch).Input.Equals("2");
+            batch = receiverSorter.ReceiveInOrder(message3, ReorderWindow).ToList();
+            Assert.Single(batch).Input.Equals("3");
+
+            Assert.Equal(0, receiverSorter.NumberBufferedRequests);
+        }
+
+        [Fact]
+        [Trait("Category", PlatformSpecificHelpers.TestCategory)]
+        public void DuplicatedElements()
+        {
+            var senderId = "A";
+            var receiverId = "B";
+
+            var senderSorter = new MessageSorter();
+
+            var message1 = Send(senderId, receiverId, "1", senderSorter, DateTime.UtcNow);
+            var message2 = Send(senderId, receiverId, "2", senderSorter, DateTime.UtcNow);
+            var message3 = Send(senderId, receiverId, "3", senderSorter, DateTime.UtcNow);
+
+            List<RequestMessage> batch;
+            MessageSorter receiverSorter;
+
+            // delivering first message last delays all messages until getting the first one
+            receiverSorter = new MessageSorter();
+            batch = receiverSorter.ReceiveInOrder(message2, ReorderWindow).ToList();
+            Assert.Empty(batch);
+            batch = receiverSorter.ReceiveInOrder(message2, ReorderWindow).ToList();
+            Assert.Empty(batch);
+            batch = receiverSorter.ReceiveInOrder(message1, ReorderWindow).ToList();
+            Assert.Collection(
+                batch,
+                first => first.Input.Equals("1"),
+                second => second.Input.Equals("2"));
+            batch = receiverSorter.ReceiveInOrder(message2, ReorderWindow).ToList();
+            Assert.Empty(batch);
+            batch = receiverSorter.ReceiveInOrder(message1, ReorderWindow).ToList();
+            Assert.Empty(batch);
+            batch = receiverSorter.ReceiveInOrder(message2, ReorderWindow).ToList();
```

Comment:
Is the second set of receiving `message1` and `message2` here necessary?

Reply:
Probably not. One could argue that there could exist some weird implementation that would fail his test, but I agree that it seems a bit excessive. I will remove it.